File: znai/src/main/java/com/twosigma/znai/structure/DocMeta.java
/*
 * Copyright 2019 TWO SIGMA OPEN SOURCE, LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.twosigma.znai.structure;

import com.twosigma.utils.JsonUtils;

import java.util.HashMap;
import java.util.Map;

public class DocMeta {
    private String id;
    private String type;
    private String title;
    private boolean previewEnabled;
    private Map<String, ?> docMetaMap;

    public DocMeta(String metaJson) {
        this(JsonUtils.deserializeAsMap(metaJson));
    }

    public DocMeta(Map<String, ?> docMetaMap) {
        this.type = docMetaMap.containsKey("type") ? docMetaMap.get("type").toString() : "no-type";
        this.title = docMetaMap.containsKey("title") ? docMetaMap.get("title").toString() : "no-title";
        this.docMetaMap = docMetaMap;
    }

    public DocMeta cloneWithNewJson(String docMetaJson) {
        DocMeta clone = new DocMeta(docMetaJson);
        clone.setId(this.id);
        clone.setPreviewEnabled(this.previewEnabled);

        return clone;
    }

    /**
     * Documentation can be standalone and hosted by itself on a server. Or you may have multiple documentations
     * that are hosted within an organization. In later case it is important to have an id.
     *
     * @return documentation id
     */
    public String getId() {
        return id;
    }

    public void setId(String id) {
        this.id = id;
    }

    public String getType() {
        return type;
    }

    public String getTitle() {
        return title;
    }

    public Map<String, Object> toMap() {
        Map<String, Object> result = new HashMap<>();
        result.put("id", id);
        result.put("previewEnabled", previewEnabled);
        result.putAll(docMetaMap);

        return result;
    }

    public boolean isPreviewEnabled() {
        return previewEnabled;
    }

    public void setPreviewEnabled(boolean previewEnabled) {
        this.previewEnabled = previewEnabled;
    }
}

File: znai-reactjs/src/structure/docMeta.ts
/*
 * Copyright 2020 znai maintainers
 * Copyright 2019 TWO SIGMA OPEN SOURCE, LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { jsonPromise } from "../utils/json";

export interface DocMeta {
  id: string;
  title: string;
  type: string;
  previewEnabled: boolean;
  slackChannel?: string;
  sendToSlackUrl?: string;
  slackActiveQuestionsUrl?: string;
  sendToSlackIncludeContentType?: boolean;
  resolveSlackQuestionUrl?: string;
  trackActivityUrl?: string;
  trackActivityIncludeContentType?: boolean;
  docStatsUrl?: string;
  useTopHeader?: boolean;
  hidePresentationTrigger?: boolean;
  support?: DocMetaSupport;
  supportMeta?: SupportMeta;
}

export interface SupportMeta {
  link: string;
  title: string;
}

export interface DocMetaSupport {
  link?: string;
  title?: string;

  urlToFetchSupportLink?(meta: DocMeta): string;
}

let docMeta: DocMeta = { id: "", title: "", type: "", previewEnabled: false };

function setDocMeta(newDocMeta: DocMeta) {
  docMeta = { ...newDocMeta };
}

function getDocMeta() {
  return { ...docMeta };
}

function mergeDocMeta(newDocMeta: Partial<DocMeta>) {
  setDocMeta({ ...docMeta, ...newDocMeta });
  return getDocMeta();
}

function isPreviewEnabled() {
  return docMeta.previewEnabled;
}

function getDocId() {
  return docMeta.id || "";
}

function isPresentationButtonVisible() {
  return !docMeta.hidePresentationTrigger;
}

let supportLinkPromise: Promise<SupportMeta> | undefined = undefined;

const defaultSupportTitle = "Support";

function getSupportLinkAndTitlePromise(): Promise<SupportMeta> {
  if (supportLinkPromise) {
    return supportLinkPromise;
  }
  const support = docMeta.support;
  if (support && support.link) {
    supportLinkPromise = new Promise((resolve) =>
      resolve({
        link: support.link!,
        title: support.title || defaultSupportTitle,
      })
    );
  } else if (support && support.urlToFetchSupportLink) {
    // @ts-ignore
    supportLinkPromise = jsonPromise<string, SupportMeta>(support.urlToFetchSupportLink(getDocMeta()))
      // @ts-ignore
      .then((supportMeta) => ({
        link: supportMeta.link,
        title: defaultSupportTitle,
      }));
  } else {
    supportLinkPromise = new Promise((resolve) => resolve({ link: "", title: "" }));
  }

  return supportLinkPromise!;
}

export {
  setDocMeta,
  mergeDocMeta,
  getDocMeta,
  isPreviewEnabled,
  getDocId,
  getSupportLinkAndTitlePromise,
  isPresentationButtonVisible,
};
